include/exec/any_sender_of.hpp

```diff
@@ -544,50 +755,119 @@ namespace exec {
       }
     };
 
-    using __unique_operation_storage = __unique_storage_t<__operation_vtable>;
+    using __immovable_operation_storage = __immovable_storage_t<__operation_vtable>;
 
     template <class _Sigs, class _Queries>
     using __receiver_ref = __mapply<__mbind_front<__q<__rec::__ref>, _Sigs>, _Queries>;
 
-    template <class _Receiver, class _Sigs, class _Queries>
+    struct __on_stop_t {
+      stdexec::in_place_stop_source& __source_;
+
+      void operator()() const noexcept {
+        __source_.request_stop();
+      }
+    };
+
+    template <class _Receiver>
     struct __operation_base {
-      STDEXEC_NO_UNIQUE_ADDRESS _Receiver __receiver_;
+      STDEXEC_NO_UNIQUE_ADDRESS _Receiver __rcvr_;
+      stdexec::in_place_stop_source __stop_source_{};
+      using __stop_callback = typename stdexec::stop_token_of_t<
+        stdexec::env_of_t<_Receiver>>::template callback_type<__on_stop_t>;
+      std::optional<__stop_callback> __on_stop_{};
+    };
+
+    template <class _Env>
+    using __env_t = __make_env_t<_Env, __with<get_stop_token_t, in_place_stop_token>>;
+
+    template <class _ReceiverId>
+    struct __stoppable_receiver {
+      using _Receiver = stdexec::__t<_ReceiverId>;
+
+      struct __t {
+        __operation_base<_Receiver>* __op_;
+
+        template <same_as<set_value_t> _SetValue, same_as<__t> _Self, class... _Args>
+          requires __callable<_SetValue, _Receiver&&, _Args...>
+        friend void tag_invoke(_SetValue, _Self&& __self, _Args&&... __args) noexcept {
+          __self.__op_->__on_stop_.reset();
+          _SetValue{}(
+            static_cast<_Receiver&&>(__self.__op_->__rcvr_), static_cast<_Args&&>(__args)...);
+        }
+
+        template <same_as<set_error_t> _SetError, same_as<__t> _Self, class _Error>
+          requires __callable<_SetError, _Receiver&&, _Error>
+        friend void tag_invoke(_SetError, _Self&& __self, _Error&& __err) noexcept {
+          __self.__op_->__on_stop_.reset();
+          _SetError{}(
+            static_cast<_Receiver&&>(__self.__op_->__rcvr_), static_cast<_Error&&>(__err));
+        }
+
+        template <same_as<set_stopped_t> _SetStopped, same_as<__t> _Self>
+          requires __callable<_SetStopped, _Receiver&&>
+        friend void tag_invoke(_SetStopped, _Self&& __self) noexcept {
+          __self.__op_->__on_stop_.reset();
+          _SetStopped{}(static_cast<_Receiver&&>(__self.__op_->__rcvr_));
+        }
+
+        template <same_as<get_env_t> _GetEnv, same_as<__t> _Self>
+        friend __env_t<env_of_t<_Receiver>> tag_invoke(_GetEnv, const _Self& __self) noexcept {
+          return __make_env(
+            get_env(__self.__op_->__rcvr_),
+            __with_(get_stop_token, __self.__op_->__stop_source_.get_token()));
+        }
+      };
```

Comment:
For future reference, my preference nowadays is to avoid nested classes. It makes diagnostics harder to grok, and it also unnecessarily adds the outer class as an associated entity that is searched for overloads. We can fix this later tho.